File: backend/app/models/order.py
import enum
from datetime import datetime
from sqlalchemy import Column, Integer, Float, String, ForeignKey, Enum, DateTime
from sqlalchemy.orm import relationship

from app.core.database import Base


class OrderStatus(str, enum.Enum):
    placed = "placed"
    accepted = "accepted"
    preparing = "preparing"
    delivery_assigned = "delivery_assigned"
    picked_up = "picked_up"
    out_for_delivery = "out_for_delivery"
    delivered = "delivered"
    cancelled = "cancelled"
    rejected = "rejected"


class PaymentMethod(str, enum.Enum):
    upi = "upi"
    cod = "cod"


class PaymentStatus(str, enum.Enum):
    pending = "pending"
    initiated = "initiated"
    paid = "paid"
    cash_received = "cash_received"
    failed = "failed"
    unverified = "unverified"


class Order(Base):
    __tablename__ = "orders"

    id = Column(Integer, primary_key=True, index=True)
    customer_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    address_id = Column(Integer, ForeignKey("addresses.id"), nullable=False)
    seller_id = Column(Integer, ForeignKey("sellers.id"), nullable=False)
    delivery_partner_id = Column(Integer, ForeignKey("delivery_partners.id"), nullable=True)

    status = Column(Enum(OrderStatus), default=OrderStatus.placed, index=True)
    subtotal = Column(Float, nullable=False)
    delivery_charge = Column(Float, default=0)
    discount = Column(Float, default=0)
    grand_total = Column(Float, nullable=False)

    payment_method = Column(Enum(PaymentMethod), nullable=False)
    payment_status = Column(Enum(PaymentStatus), default=PaymentStatus.pending, index=True)

    created_at = Column(DateTime, default=datetime.utcnow)

    items = relationship("OrderItem", back_populates="order", cascade="all, delete-orphan")
    status_history = relationship("OrderStatusHistory", back_populates="order", cascade="all, delete-orphan")

    # These were referenced (order.customer, order.seller, order.address)
    # in Phase 6 code (delivery.py, invoice endpoint) but never actually
    # declared — that would have raised AttributeError at runtime the
    # first time those endpoints ran. Fixed here as part of Phase 6.5
    # hardening; no DB schema change, just exposing the existing FKs.
    customer = relationship("User", foreign_keys=[customer_id])
    seller = relationship("Seller", foreign_keys=[seller_id])
    address = relationship("Address", foreign_keys=[address_id])
    delivery_partner = relationship("DeliveryPartner", foreign_keys=[delivery_partner_id])


class OrderItem(Base):
    __tablename__ = "order_items"

    id = Column(Integer, primary_key=True, index=True)
    order_id = Column(Integer, ForeignKey("orders.id"), nullable=False)
    product_id = Column(Integer, ForeignKey("products.id"), nullable=False)
    quantity = Column(Float, nullable=False)
    unit_price = Column(Float, nullable=False)
    amount = Column(Float, nullable=False)

    order = relationship("Order", back_populates="items")
    product = relationship("Product")


class OrderStatusHistory(Base):
    __tablename__ = "order_status_history"

    id = Column(Integer, primary_key=True, index=True)
    order_id = Column(Integer, ForeignKey("orders.id"), nullable=False)
    status = Column(Enum(OrderStatus), nullable=False)
    changed_by_user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    timestamp = Column(DateTime, default=datetime.utcnow)

    order = relationship("Order", back_populates="status_history")

File: seller_app/lib/models/order.dart
class OrderSummary {
  final int id;
  final String status;
  final String paymentMethod;
  final String paymentStatus;
  final double grandTotal;
  final DateTime? createdAt;
  final int itemCount;

  OrderSummary({
    required this.id,
    required this.status,
    required this.paymentMethod,
    required this.paymentStatus,
    required this.grandTotal,
    this.createdAt,
    required this.itemCount,
  });

  factory OrderSummary.fromJson(Map<String, dynamic> json) {
    double num(dynamic v) => v == null ? 0 : (v is int || v is double ? v.toDouble() : double.tryParse(v.toString()) ?? 0);
    return OrderSummary(
      id: json['id'],
      status: json['status']?.toString() ?? 'placed',
      paymentMethod: json['payment_method']?.toString() ?? 'cod',
      paymentStatus: json['payment_status']?.toString() ?? 'pending',
      grandTotal: num(json['grand_total']),
      createdAt: json['created_at'] != null ? DateTime.tryParse(json['created_at'].toString()) : null,
      itemCount: json['item_count'] ?? 0,
    );
  }
}

class OrderItemLine {
  final int productId;
  final String name;
  final double quantity;
  final double unitPrice;
  final double amount;

  OrderItemLine({required this.productId, required this.name, required this.quantity, required this.unitPrice, required this.amount});

  factory OrderItemLine.fromJson(Map<String, dynamic> json) {
    double num(dynamic v) => v == null ? 0 : (v is int || v is double ? v.toDouble() : double.tryParse(v.toString()) ?? 0);
    return OrderItemLine(
      productId: json['product_id'],
      name: json['name']?.toString() ?? '',
      quantity: num(json['quantity']),
      unitPrice: num(json['unit_price']),
      amount: num(json['amount']),
    );
  }
}

class StatusHistoryEntry {
  final String status;
  final DateTime? timestamp;
  StatusHistoryEntry({required this.status, this.timestamp});
  factory StatusHistoryEntry.fromJson(Map<String, dynamic> json) => StatusHistoryEntry(
        status: json['status']?.toString() ?? '',
        timestamp: json['timestamp'] != null ? DateTime.tryParse(json['timestamp'].toString()) : null,
      );
}

class CustomerInfo {
  final String? name;
  final String? phone;
  CustomerInfo({this.name, this.phone});
  factory CustomerInfo.fromJson(Map<String, dynamic>? json) =>
      CustomerInfo(name: json?['name']?.toString(), phone: json?['phone']?.toString());
}

class DeliveryAddressInfo {
  final String villageTown;
  final String house;
  final String? landmark;
  final String pincode;
  final String mobile;
  DeliveryAddressInfo({required this.villageTown, required this.house, this.landmark, required this.pincode, required this.mobile});
  factory DeliveryAddressInfo.fromJson(Map<String, dynamic> json) => DeliveryAddressInfo(
        villageTown: json['village_town']?.toString() ?? '',
        house: json['house']?.toString() ?? '',
        landmark: json['landmark']?.toString(),
        pincode: json['pincode']?.toString() ?? '',
        mobile: json['mobile']?.toString() ?? '',
      );
}

class OrderDetail {
  final int id;
  final String status;
  final String paymentMethod;
  final String paymentStatus;
  final double subtotal;
  final double deliveryCharge;
  final double discount;
  final double grandTotal;
  final List<OrderItemLine> items;
  final List<StatusHistoryEntry> statusHistory;
  final CustomerInfo customer;
  final DeliveryAddressInfo? address;

  OrderDetail({
    required this.id,
    required this.status,
    required this.paymentMethod,
    required this.paymentStatus,
    required this.subtotal,
    required this.deliveryCharge,
    required this.discount,
    required this.grandTotal,
    required this.items,
    required this.statusHistory,
    required this.customer,
    this.address,
  });

  factory OrderDetail.fromJson(Map<String, dynamic> json) {
    double num(dynamic v) => v == null ? 0 : (v is int || v is double ? v.toDouble() : double.tryParse(v.toString()) ?? 0);
    return OrderDetail(
      id: json['id'],
      status: json['status']?.toString() ?? 'placed',
      paymentMethod: json['payment_method']?.toString() ?? 'cod',
      paymentStatus: json['payment_status']?.toString() ?? 'pending',
      subtotal: num(json['subtotal']),
      deliveryCharge: num(json['delivery_charge']),
      discount: num(json['discount']),
      grandTotal: num(json['grand_total']),
      items: (json['items'] as List? ?? []).map((e) => OrderItemLine.fromJson(e)).toList(),
      statusHistory: (json['status_history'] as List? ?? []).map((e) => StatusHistoryEntry.fromJson(e)).toList(),
      customer: CustomerInfo.fromJson(json['customer']),
      address: json['address'] != null ? DeliveryAddressInfo.fromJson(json['address']) : null,
    );
  }
}
